docs/project-architecture.md

```diff
@@ -0,0 +1,47 @@
+# Project Architecture
+
+This is a high-level overview of the Visual Studio Code Extension. The intended audience is someone interested in contributing the extension.
+
+## Background
+
+Visual Studio Code provides an [API](https://code.visualstudio.com/api/language-extensions/overview) for implementing language-specific editing features in an extension. In VS Code, a language editor extension has two main components:
+
+- a **language client** which is a normal VS Code extension that listens to different editor events and sends the payloads of these events to a language server
+- a **language server** which is performs language analysis on the code written within VSCode (this runs in a separate process for performance reasons)
+
+The language client and server communicate via the Language Server Protocol. The [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) defines a standardized communication protocol between an editor (in our case Visual Studio Code) and a language server that implements features like autocomplete and syntax checking.
+
+![LSP Illustration](./images/lsp-illustration.png)
+
+By making the extension LSP compliant we have set the extension up to integrate with other text editors in the future.
```

Comment:
The extension will not integrate with other editors, but the separate language server implementation will allow us to integrate into other editors.